num_enum_derive/src/lib.rs

```diff
@@ -422,75 +441,116 @@ impl Parse for EnumInfo {
-                                    die!(&alternative_values[i] => format!("'{}' in the alternative values is already attributed to a previous variant", val));
+                    // Search if those discriminant_int_val_set where already attributed.
+                    // (discriminant_int_val_set is BTreeSet, and iter().next_back() is the is the maximum in the set.)
+                    if let Some(last_upper_val) = discriminant_int_val_set.iter().next_back() {
+                        if sorted_alternate_int_values.first().unwrap() <= last_upper_val {
+                            for (i, val) in sorted_alternate_int_values.iter().enumerate() {
+                                if discriminant_int_val_set.contains(val) {
+                                    die!(&raw_alternative_values[i] => format!("'{}' in the alternative values is already attributed to a previous variant", val));
                                 }
                             }
                         }
                     }
 
                     // Reconstruct the alternative_values vec of Expr but sorted.
-                    alternative_values = alt_val_sorted
+                    raw_alternative_values = sorted_alternate_int_values
                         .iter()
                         .map(|val| literal(val.to_owned()))
                         .collect();
 
                     // Add the alternative values to the the set to keep track.
-                    val_set.extend(alt_val_sorted);
+                    discriminant_int_val_set.extend(sorted_alternate_int_values);
                 }
 
                 // Add the current discriminant to the the set to keep track.
-                let newly_inserted = val_set.insert(canonical_value_int);
-                debug_assert!(newly_inserted);
+                if let DiscriminantValue::Literal(canonical_value_int) = discriminant_value {
+                    discriminant_int_val_set.insert(canonical_value_int);
+                }
 
                 variants.push(VariantInfo {
                     ident,
                     attr_spans,
                     is_default,
                     is_catch_all,
-                    canonical_value,
-                    alternative_values,
+                    canonical_value: discriminant,
+                    alternative_values: raw_alternative_values,
                 });
 
                 // Get the next value for the discriminant.
-                next_discriminant = literal(canonical_value_int + 1);
+                next_discriminant = match discriminant_value {
+                    DiscriminantValue::Literal(int_value) => {
+                        if int_value >= -1 {
+                            literal(int_value + 1)
+                        } else {
+                            let value = literal((int_value + 1).abs());
+                            parse_quote! {
+                                -#value
+                            }
+                        }
```

Comment:
Maybe a `literal` can still be negative if hand-crafted? I suspect the AST to be ambiguous here, and even if by default it parses a `-<literal>` as a `-` operation on a `<literal>`, I think you can also use a `<literal>` expression with a hand-crafted negative value inside it. But I have not tested it.

Also, I'm a bit nervous about implicitly relying on wrapping arithmetic; I think this was an oversight of previous code of ours, and I think it could be a good chance to fix it:

```suggestion
                        let next_int = int_value.wrapping_add(1);
                        if next_int >= 0 {
                            literal(next_int)
                        } else {
                            let value = literal(next_int.abs());
                            parse_quote! {
                                -#value
                            }
                        }
```